src/Components/Aspire.Azure.Npgsql/AspireAzureNpgsqlExtensions.cs

```diff
@@ -0,0 +1,215 @@
+    /// <exception cref="ArgumentNullException">Thrown when <paramref name="builder"/> or <paramref name="name"/> is null.</exception>
+    /// <exception cref="ArgumentException">Thrown if mandatory <paramref name="name"/> is empty.</exception>
+    /// <exception cref="InvalidOperationException">Thrown when mandatory <see cref="NpgsqlSettings.ConnectionString"/> is not provided or the <see cref="AzureNpgsqlSettings.Credential"/> is invalid.</exception>
+    public static void AddKeyedAzureNpgsqlDataSource(this IHostApplicationBuilder builder, string name, Action<AzureNpgsqlSettings>? configureSettings = null, Action<NpgsqlDataSourceBuilder>? configureDataSourceBuilder = null)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+        ArgumentException.ThrowIfNullOrEmpty(name);
+
+        AzureNpgsqlSettings? azureSettings = null;
+
+        builder.AddKeyedNpgsqlDataSource(name, settings => azureSettings = ConfigureSettings(configureSettings, settings), dataSourceBuilder =>
+        {
+            Debug.Assert(azureSettings != null);
+
+            ConfigureDataSourceBuilder(azureSettings, dataSourceBuilder);
+
+            configureDataSourceBuilder?.Invoke(dataSourceBuilder);
+        });
+    }
+
+    private static AzureNpgsqlSettings ConfigureSettings(Action<AzureNpgsqlSettings>? userConfigureSettings, NpgsqlSettings settings)
+    {
+        var azureSettings = new AzureNpgsqlSettings();
+
+        // Copy the values updated by Npgsql integration.
+        CopySettings(settings, azureSettings);
+
+        // Invoke the Aspire configuration.
+        userConfigureSettings?.Invoke(azureSettings);
+
+        // Copy to the Npgsql integration settings as it needs to get any values set in userConfigureSettings.
+        CopySettings(azureSettings, settings);
+
+        return azureSettings;
+    }
+
+    private static void CopySettings(NpgsqlSettings source, NpgsqlSettings destination)
+    {
+        destination.ConnectionString = source.ConnectionString;
+        destination.DisableHealthChecks = source.DisableHealthChecks;
+        destination.DisableMetrics = source.DisableMetrics;
+        destination.DisableTracing = source.DisableTracing;
+    }
+
+    private static void ConfigureDataSourceBuilder(AzureNpgsqlSettings settings, NpgsqlDataSourceBuilder dataSourceBuilder)
+    {
+        // The connection string required the username to be provided. Since it will depend on the Managed Identity that is used
+        // we attempt to get the username from the access token.
+
+        var credential = settings.Credential ?? new DefaultAzureCredential();
+
+        if (string.IsNullOrEmpty(dataSourceBuilder.ConnectionStringBuilder.Username))
+        {
+            // ensure to use the management scope, so the token contains user names for all managed identity types - e.g. user and service principal
+            var token = credential.GetToken(new([AzureManagementScope]), default);
```

Comment:
Commented on this line in another discussion thread:

> One case to consider: is there instances where a user would have permissions for the AzureDatabaseForPostgresSqlScope, and not AzureManagementScope ?

https://github.com/dotnet/aspire/pull/8230#discussion_r2012103423